src/worker/tasks/balance-subscription-worker.ts

```diff
@@ -0,0 +1,136 @@
+import { type Job, type Processor, Worker } from "bullmq";
+import { getAllBalanceSubscriptions } from "../../shared/db/balance-subscriptions/get-all-balance-subscriptions";
+import { getConfig } from "../../shared/utils/cache/get-config";
+import { logger } from "../../shared/utils/logger";
+import { redis } from "../../shared/utils/redis/redis";
+import { BalanceSubscriptionQueue } from "../queues/balance-subscription-queue";
+import { logWorkerExceptions } from "../queues/queues";
+import { SendWebhookQueue } from "../queues/send-webhook-queue";
+import {
+  WebhooksEventTypes,
+  type BalanceSubscriptionWebhookParams,
+} from "../../shared/schemas/webhooks";
+import { parseBalanceSubscriptionConfig } from "../../shared/schemas/balance-subscription-config";
+import { getChain } from "../../shared/utils/chain";
+import { eth_getBalance, getContract, getRpcClient } from "thirdweb";
+import { thirdwebClient } from "../../shared/utils/sdk";
+import { balanceOf } from "thirdweb/extensions/erc20";
+import { maxUint256 } from "thirdweb/utils";
+
+// Split array into chunks of specified size
+function chunk<T>(arr: T[], size: number): T[][] {
+  return Array.from({ length: Math.ceil(arr.length / size) }, (_, i) =>
+    arr.slice(i * size, i * size + size),
+  );
+}
+
+// Must be explicitly called for the worker to run on this host.
+export const initBalanceSubscriptionWorker = async () => {
+  const config = await getConfig();
+  const cronPattern =
+    config.balanceSubscriptionsCronSchedule || "*/2 * * * * *"; // Default to every 30 seconds
+
+  logger({
+    service: "worker",
+    level: "info",
+    message: `Initializing balance subscription worker with cron pattern: ${cronPattern}`,
+  });
+
+  BalanceSubscriptionQueue.q.add("cron", "", {
+    repeat: { pattern: cronPattern },
+    jobId: "balance-subscription-cron",
+  });
+
+  const _worker = new Worker(BalanceSubscriptionQueue.q.name, handler, {
+    connection: redis,
+    concurrency: 1,
+  });
+  logWorkerExceptions(_worker);
+};
+
+/**
+ * Process all balance subscriptions and notify webhooks of changes.
+ */
+const handler: Processor<string, void, string> = async (job: Job<string>) => {
+  // Get all active balance subscriptions
+  const subscriptions = await getAllBalanceSubscriptions();
+  if (subscriptions.length === 0) {
+    return;
+  }
+
+  // Process in batches of 50
+  const batches = chunk(subscriptions, 50);
+  for (const batch of batches) {
+    await Promise.all(
+      batch.map(async (subscription) => {
+        try {
+          // Get the current balance
+          let currentBalance: bigint;
+          const chain = await getChain(Number.parseInt(subscription.chainId));
+          if (subscription.contractAddress) {
+            const contract = getContract({
+              address: subscription.contractAddress,
+              chain: chain,
+              client: thirdwebClient,
+            });
+
+            currentBalance = await balanceOf({
```

Comment:
wait what is contract here? this is to get a erc20 balance, that's not what you want here

Reply:
Yes you can subscribe to either erc20 balance or native balance changes